enhancements/machine-api/control-plane-set.md

```diff
@@ -0,0 +1,585 @@
+
+```go
+type ControlPlaneSetSpec struct {
+    // Replicas defines how many Control Plane Machines should be
+    // created by this ControlPlaneSet.
+    // This field is immutable and cannot be changed after cluster
+    // installation.
+    // The ControlPlaneSet only operates with 3 or 5 node control planes,
+    // 3 and 5 are the only valid values for this field.
+    // +kubebuilder:validation:Enum:=3;5
+    // +kubebuilder:default:=3
+    // +kubebuilder:validation:Required
+    Replicas *int32 `json:"replicas"`
+
+    // Strategy defines how the ControlPlaneSet will update
+    // Machines when it detects a change to the ProviderSpec.
+    // +optional
+    Strategy ControlPlaneSetStrategy `json:"strategy,omitempty"`
+
+    // FailureDomains is the list of failure domains or
+    // availability zones in which the ControlPlaneSet should
+    // balance the Control Plane Machines.
+    // This will beinjected into the ProviderSpec in the
+    // appropriate location for the particular provider.
+    // +kubebuilder:validation:MinLength:=1
+    // +kubebuilder:validation:Required
+    FailureDomains []string `json:"failureDomains"`
+
+    // Label selector for Machines. Existing Machines selected by this
+    // selector will be the ones affected by this ControlPlaneSet.
+    // It must match the template's labels.
+    // This field is considered immutable after creation of the resource.
+    // +kubebuilder:validation:Required
+    Selector metav1.LabelSelector `json:"selector"`
+
+    // Template describes the Control Plane Machines that will be created
+    // by this ControlPlaneSet.
+    // +kubebuilder:validation:Required
+    Template ControlPlaneSetTemplate `json:"template"`
+}
+
+type ControlPlaneSetTemplate struct {
+    // ObjectMeta is the standard object metadata
+    // More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+    // Labels are required to match the ControlPlaneSet selector.
+    // +kubebuilder:validation:Required
+    ObjectMeta metav1.ObjectMeta `json:"metadata"`
+
+    // ProviderSpec contains the platform specific details
+    // for creating the Control Plane Machines.
+    // It should be complete apart from the platform specific
+    // failure domain field. This will be overriden when the
+    // Machines are created based on the FailureDomains field.
+    // +kubebuilder:validation:Required
+    ProviderSpec *runtime.RawExtension `json:"providerSpec"`
+}
+
+type ControlPlaneSetStrategy struct {
+    // Type defines the type of update strategy that should be
+    // used when updating Machines owned by the ControlPlaneSet.
+    // Valid values are "RollingUpdate", "Recreate" and "OnDelete".
+    // The current default value is "RollingUpdate".
+    // +kubebuilder:default:="RollingUpdate"
+    // +kubebuilder:validation:Enum:="RollingUpdate";"Recreate";"OnDelete"
+    // +optional
+    Type ControlPlaneSetStrategyType `json:"type,omitempty"`
```

Comment:
From a UX pov I see two upgrade types that cover needs in existing scenarios - replace and in-place. Now, you can define different strategies to dictate the upgrade workflow in each scenario depending on how much control you need e.g rolling, ondelete...
These are the semantics we expose in hypershift:
UpgradeType (replace/in-place) 1 - n UpgradeStrategy (RollingUpdate, OnDelete...)

Reply:
If I understand this correctly, the suggestion is about the way to modify the hosts? So replace would mean create a new host and in-place would mean updating the host on the infrastructure provider in-place.

Is in-place something that HyperShift is planning to support? I was under the impression that both MAPI and CAPI were not going to do in-place reconfiguration and that they were considered to be immutable infrastructure projects?

Reply:
We do support in place upgrades in OCP regardless what API is backing a Node, and we do in hypershift via NodePools.

Reply:
Do you mean from a hardware perspective or from a software perspective? I appreciate that we update the software on our hosts, but I don't think Machine API at least has a way to reconcile the hardware aspects (eg if a user were to change an instance type we can't update that on the fly). Are you suggesting this can be done in HyperShift? If so, I need to learn more about this, perhaps we can sync

Reply:
Had a sync with Alberto offline to understand what is going on here, this is a software level feature which we do not intend to support initially within CPMS.
The API design does not preclude us from adding this in the future, but the initial requirements dictate hardware replacement, so to limit scope, having a distinction between hardware and software replacements isn't required